pkg/diagnostics/cluster/registry.go

```diff
@@ -0,0 +1,281 @@
+	if err != nil {
+		r.Errorf("clRegListPods", err, "Finding pods for '%s' service failed. This should never happen. Error: (%T) %[2]v", registryName, err)
+		return runningPods
+	} else if len(pods.Items) < 1 {
+		r.Errorf("clRegNoPods", nil, clRegNoPods, registryName)
+		return runningPods
+	} else if len(pods.Items) > 1 {
+		// multiple registry pods using EmptyDir will be inconsistent
+		for _, volume := range pods.Items[0].Spec.Volumes {
+			if volume.Name == registryVolume && volume.EmptyDir != nil {
+				r.Errorf("clRegMultiPods", nil, clRegMultiPods, registryName)
+				break
+			}
+		}
+	}
+	for _, pod := range pods.Items {
+		r.Debugf("clRegPodFound", "Found %s pod with name %s", registryName, pod.ObjectMeta.Name)
+		if pod.Status.Phase != kapi.PodRunning {
+			r.Warnf("clRegPodDown", nil, clRegPodDown, pod.ObjectMeta.Name, registryName)
+		} else {
+			runningPods = append(runningPods, &pod)
+			// Check the logs for that pod for common issues (credentials, DNS resolution failure)
+			d.checkRegistryLogs(&pod, r)
+		}
+	}
+	return runningPods
+}
+
+func (d *ClusterRegistry) checkRegistryLogs(pod *kapi.Pod, r types.DiagnosticResult) {
+	// pull out logs from the pod
+	readCloser, err := d.KubeClient.RESTClient.Get().
+		Namespace("default").Name(pod.ObjectMeta.Name).
+		Resource("pods").SubResource("log").
+		Param("follow", "false").
+		Param("container", pod.Spec.Containers[0].Name).
+		Stream()
+	if err != nil {
+		r.Warnt("clRegPodLog", nil, clRegPodLog, log.Hash{
+			"error":        fmt.Sprintf("(%T) %[1]v", err),
+			"podName":      pod.ObjectMeta.Name,
+			"registryName": registryName,
+		})
+		return
+	}
+	defer readCloser.Close()
+
+	scanner := bufio.NewScanner(readCloser)
+	for scanner.Scan() {
+		logLine := scanner.Text()
+		if regexp.MustCompile(`level=error msg="client error: Post http(\S+)/subjectaccessreviews`).MatchString(logLine) {
+			r.Errort("clRegPodConn", nil, clRegPodConn, log.Hash{
+				"log":          logLine,
+				"podName":      pod.ObjectMeta.Name,
+				"registryName": registryName,
+			})
+			break
+		}
+	}
+}
+
+func (d *ClusterRegistry) checkRegistryEndpoints(pods []*kapi.Pod, r types.DiagnosticResult) bool {
+	endPoint, err := d.KubeClient.Endpoints(kapi.NamespaceDefault).Get(registryName)
+	if err != nil {
+		r.Errorf("clRegGetEP", err, `Finding endpoints for "%s" service failed. This should never happen. Error: (%[2]T) %[2]v`, registryName, err)
+		return false
+	}
+	numEP := 0
+	for _, subs := range endPoint.Subsets {
+		numEP += len(subs.Addresses)
+	}
+	if numEP != len(pods) {
+		r.Warnt("clRegNoEP", nil, clRegNoEP, log.Hash{"registryName": registryName, "numPods": len(pods), "numEP": numEP})
+		return false
+	}
+	return true
+}
+
+func (d *ClusterRegistry) verifyRegistryImageStream(service *kapi.Service, r types.DiagnosticResult) {
+	imgStream, err := d.OsClient.ImageStreams(kapi.NamespaceDefault).Create(&osapi.ImageStream{ObjectMeta: kapi.ObjectMeta{GenerateName: "diagnostic-test"}})
```

Comment:
I did have some trepidation about this, but I can't think of any other way to test for this condition from a client, and it's really puzzling when it happens. Note that it's clearly communicated if the thing generated fails to be deleted for some reason. I'm open to suggestions...?